Add design doc for extending Envoy Gateway (#1074)
envoyproxy/gateway: docs/latest/design/extending-envoy-gateway.md

Kubernetes Gateway API conformance. This system focuses on the “common foundation for vendors” component by introducing a way for vendors to extend Envoy Gateway.

To meaningfully extend Envoy Gateway and provide additional features, Extensions need to be able to introduce their own custom resources and have a high level of control
over the configuration generated by Envoy Gateway. Simply applying some static xDS configuration patches or relying on the existing Gateway API resources are both insufficient on their own
as means to add larger features that require dynamic user-configuration.

As an example, an extension developer may wish to provide their own out of the box authentication filters that require configuration from the end-user. This is a scenario where the ability to introduce
custom resources and attach them to [HTTPRoute][]s as an [ExtensionRef][] is necessary. Providing the same feature through a series of xDS patch resources would be too cumbersome for many end-users that want to avoid
that level of complexity when managing their clusters.  

## Goals

- Provide a foundation for extending the Envoy Gateway control plane

> - Allow Extension Developers to introudce their own cusom resources for extending the Gateway-API via [ExtensionRefs][], [policyAttachments][] (future) and [backendRefs][] (future).

Reviewer: typo: introudce -> introduce

Reviewer 2: why this pass spellcheck CI?

> - Allow Extension Developers to introudce their own cusom resources for extending the Gateway-API via [ExtensionRefs][], [policyAttachments][] (future) and [backendRefs][] (future).

Reviewer 3: ```suggestion
- Allow Extension Developers to introduce their own custom resources for extending the Gateway-API via [ExtensionRefs][], [policyAttachments][] (future) and [backendRefs][] (future).
```

- Extension developers should **NOT** have to maintain a custom fork of Envoy Gateway
- Provide a system for extending Envoy Gateway which allows extension projects to ship updates independent of Envoy Gateway's release schedule
- Extensions can hook into both the Envoy configuration pipeline and the infra manager pipeline
- Modify the generated Envoy xDS and Envoy Gateway infra config (either directly via an xDS hook or earlier in the pipeline or by dynamically adding resources)
- Setup a foundation for the initial iteration of Extending Envoy Gateway

## Non-Goals

> - The initial deisgn does not capture every hook that Envoy Gateway will eventually support.

Reviewer 3: ```suggestion
- The initial design does not capture every hook that Envoy Gateway will eventually support.
```

- Extend [Gateway API Policy Attachments][]. At some point these will be addressed using this extension system, but the initial implementation omits these.
- Guarantee that multiple extensions will work together without conflicts

## Overview

Envoy Gateway can be extended by vendors by means of an extension server developed by the vendor and deployed alongside Envoy Gateway.

> An extension server can make use of one or more pre/post hooks inside Envoy Gateway before and after it's major components (translator, etc.) to allow the extension to modify the data going into or coming out of these components.

Reviewer 3: ```suggestion
An extension server can make use of one or more pre/post hooks inside Envoy Gateway before and after its major components (translator, etc.) to allow the extension to modify the data going into or coming out of these components.
```

> An extension can be created external to Envoy Gateway as it's own Kubernetes deployment or loaded as a sidecar. gRPC is used for the calls between Envoy Gateway and an extension. In the hook call, Envoy Gateway sends data as well 

Reviewer 3: ```suggestion
An extension can be created external to Envoy Gateway as its own Kubernetes deployment or loaded as a sidecar. gRPC is used for the calls between Envoy Gateway and an extension. In the hook call, Envoy Gateway sends data as well 
```

          - OAuth2Filter
    service:
      host: my-extension.example
      port: 443
      tls:
        type: secret
        secret:
            name: my-extension
            namespace: example
```

This configuration is required to be provided at bootstrap and modifying the registered extensions during runtime is not currently supported.
Envoy Gateway will keep track of the registered extensions and their `apiGroups` and `kinds` when processing Gateway API resources.

## Extending Gateway API and the Data Plane

Envoy Gateway manages [Envoy][] deployments, which act as the data plane that handles actual user traffic. Users configure the data plane using the K8s Gateway API resources which Envoy 
Gateway converts into [Envoy specific configuration (xDS)][] to send over to Envoy.

Gateway API offers [ExtensionRef filters][] and [Policy Attachments][] as extension points for implementers to use. Envoy Gateway extends the Gateway API using these extension points to provide support for [rate limiting][] 
and [authentication][]. The initial design of Envoy Gateway extensions will primarily focus on ExtensionRef filters, where users can reference a custom resource as an HTTP Filter.

> When Envoy Gateway encounters an [HTTPRoute][] or [GRPCRoute][] that has an `ExtensionRef` `filter` with an `apiGroup` and `kind` that Envoy Gateway does not supoport, it will first

Reviewer 3: ```suggestion
When Envoy Gateway encounters an [HTTPRoute][] or [GRPCRoute][] that has an `ExtensionRef` `filter` with an `apiGroup` and `kind` that Envoy Gateway does not support, it will first
```

check the list of registered extensions to determine if any of the registered extensions support the referenced object before considering it a configuration error.

This allows users to be able to reference additional filters provided by extensions to Envoy Gateway, in their `HTTPRoute`s / `GRPCRoute`s:

```yaml
apiVersion: example.myextension.io/v1alpha1
kind: OAuth2Filter
metadata:
  name: oauth2-filter
spec:
  ...

---
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: example
spec:
  parentRefs:
  - name: eg
  hostnames:
  - www.example.com
  rules:
  - clientSelectors:
    - path:
        type: PathPrefix
        value: /
    filters:
    - type: ExtensionRef
      extensionRef:
        group: example.myextension.io
        kind: OAuth2Filter
        name: oauth2-filter
    backendRefs:
    - name: backend
      port: 3000
```

In order to enable the translation of new resources introduced by extensions, Envoy Gateway provides hook points within the translation pipeline, where it calls out to extension services registered in the [EnvoyGateway config][] 

> if they specify an `apiGroup` that matches the `group` of an `ExtensionRef` filter. The extension will then be able to modify the xDS that Envoy Gateway generated and send back the 

Reviewer 3: Does it mean that the extension will work in the xDS Translator layer?
Envoy gateway will send the xDS IR to the extension and the extension send the xDS patch back to the Envoy gateway.

Author: Correct. For the initial implementation of this design, there will likely only be a single hook point where an extension has control over the xDS resources right before Envoy Gateway updates them in the xDS server to be consumed by Envoy Proxy. Still, the framework of the extensions system is meant to be a foundation where more fine-grained hooks can be introduced into the future for use cases where the extension does not need such broad control.

Reviewer 3: Thanks for your reply.

To power these use-cases, Envoy Gateway provides additional hook entrypoints to the Infra Manager which is responsible for dynamically creating the
instances of Envoy Proxy. These hook points empower extensions to modify the Envoy Proxy deployment and service dynamically.

In the example above where an extension developer wants to create a new `ExtensionRef` filter that provides authentication, they would likely deploy their own authorization service into the cluster
and make use of the extension hooks to configure Envoy to use that service to authenticate requests. By also using extension points in the control plane, the extension can dynamically
inject the authorization service as a sidecar to the Envoy deployments created by Envoy Gateway rather than deploy the authorization service in the Kubernetes cluster separately. Doing so would help to minimize the 
latency between Envoy and the authentication service and also allows the end user to spend less time worrying about how things are deployed in the cluster.

![Control Plane Extension Example](../images/extension-auth-example.png)

## Known Challenges

> Extending Envoy Gateway by using an external extension server which makes use of hook points in Envoy Gateway does come with some know issues that will have to be worked around.

Reviewer 3: ```suggestion
Extending Envoy Gateway by using an external extension server which makes use of hook points in Envoy Gateway does come with some known issues that will have to be worked around.
```

> - Extension server handling it's own state such as in the scenario where an extension establishes watches its own CRDs in the cluster can lead to a scenario where

Reviewer 4: Is it possible to let Envoy Gateway wrap the custom CRD resources in the context when calling the extenstion server hooks?

> - Extension server handling it's own state such as in the scenario where an extension establishes watches its own CRDs in the cluster can lead to a scenario where

Reviewer 3: ```suggestion
- Extension server handling its own state such as in the scenario where an extension establishes watches its own CRDs in the cluster can lead to a scenario where
```

ensuring that the extension and Envoy Gateway are in-sync becomes a challenge.

>   - A potential workaround to this issue would be to have the extension use hook points in Envoy Gateway to replace Envoy Gateway's Kubernetes provider (the component that watches resources such as the 

Reviewer 4: Dose this mean that Envoy Gateway watches the custom CRD resources and wrap the custom CRD resources in the RPC context when calling the extenstion server webhooks?

Does this mean that the Envoy Gateway watches the CRD resources for extensions and includes them in the RPC context when calling the extension server webhooks?

Reviewer 4: I think it may be too much to ask extensions to take over all provider responsibilities. Is it feasible for an extension to only monitor its own resources? If consistency is a concern, I feel that the approach of 'the provider listens to all resources and passes them back to the extension as needed' sounds more reasonable.

An extension can be created external to Envoy Gateway as its own Kubernetes deployment or loaded as a sidecar. gRPC is used for the calls between Envoy Gateway and an extension. In the hook call, Envoy Gateway sends data as well 
as context information to the extension and expects a reply with a modified version of the data that was sent to the extension. Since extensions fundamentally alter the logic and data that Envoy Gateway provides, Extension projects assume responsibility for any bugs and issues 
they create as a direct result of their modification of Envoy Gateway.

## Diagram

![Architecture](../images/extension-example.png)

## Registering Extensions in Envoy Gateway

Information about the extension that Envoy Gateway needs to load is configured in the Envoy Gateway config. The extension must specify:

- Which API groups the extension supports
- The `Kind`s supported under each API group
- Extension service connection info (host, port, TLS settings, etc.)

An example configuration:

```yaml
apiVersion: config.gateway.envoyproxy.io/v1alpha1
kind: EnvoyGateway
extension:

>   apiGroups:

Reviewer 5: ```suggestion
  extensionRefs:
```

>   - apiGroup: example.myextension.io

Reviewer 5: ```suggestion
  - group: example.myextension.io
```

>     apiVersion: v2

Reviewer 5: ```suggestion
    version: v2
```

>     kinds:

Reviewer 5: reusing similar language as https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.LocalObjectReference

    - OAuth2Filter
  service:
    host: my-extension.example
    port: 443

>     tls:

Reviewer 5: can this look similar to https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.GatewayTLSConfig ?

      type: secret
      secret:
        name: my-extension
        namespace: example
```

This configuration is required to be provided at bootstrap and modifying the registered extension during runtime is not currently supported.
Envoy Gateway will keep track of the registered extension and its `apiGroups` and `kinds` when processing Gateway API resources.

## Extending Gateway API and the Data Plane

Envoy Gateway manages [Envoy][] deployments, which act as the data plane that handles actual user traffic. Users configure the data plane using the K8s Gateway API resources which Envoy 
Gateway converts into [Envoy specific configuration (xDS)][] to send over to Envoy.

> Gateway API offers [ExtensionRef filters][] and [Policy Attachments][] as extension points for implementers to use. Envoy Gateway extends the Gateway API using these extension points to provide support for [rate limiting][] 

Reviewer 5: how can we better frame this to highlight 2 types of extensions
* Native to the project
* Extending the control plane by the vendor/extension dev

registered extension configured. When communicating with an extension, Envoy Gateway sends these Unstructured resources over to the extension. This eliminates the need for the extension to create its own watches which would have a strong chance of creating race conditions and reconciliation loops when resources change. When an extension receives the Unstructured resources from Envoy Gateway it can perform its own type validation on them. Currently we make the simplifying assumption that the registered extension's `Kinds` are filters referenced by `extensionRef` in `HTTPRouteFilter`s . Support for Policy attachments will be introduced at a later time.

## Control Plane Extensions

To provide extensions the flexibility to do many things such as introduce a new authentication filter that introduces
its own new [external authorization][] server, it is also necessary to be able to extend the Control plane.

To power these use-cases, Envoy Gateway provides additional hook entrypoints to the Infra Manager which is responsible for dynamically creating the
instances of Envoy Proxy. These hook points empower an extension to modify the Envoy Proxy deployment and service dynamically.

> In the example above where an extension developer wants to create a new `ExtensionRef` filter that provides authentication, they would likely deploy their own authorization service into the cluster

Reviewer 5: is extending infra manager part of the proposal as well ?

focuses solely on the modification of xDS resources.

### Route Modification Hook

The Route Hook provides a way for an extension to modify a route generated by Envoy Gateway before it is finalized.
Doing so allows the extension to configure/modify route fields configured by Envoy Gateway and also to configure the
Route's TypedPerFilterConfig which may be desirable to do things such as pass settings and information to
ext_authz filters. The Route Hook also passes a list of Unstructured data for the externalRefs owned by the extension on the HTTPRoute that created this xDS route. The Route Hook will only be executed if an extension is loaded and only on Routes that were generated from an HTTPRoute that uses extension resources as externalRef filters.

```go
// RouteModifyRequest sends over the Route that was generated by Envoy Gateway along with some context about how it was generated
// to assist the extension in making any necessary changes to the route.
message RouteModifyRequest {

>     envoy.config.route.v3.Route route = 1;

Reviewer 5: what links a `route` to a `route_context` ?

    RouteExtensionContext route_context = 2;
    repeated ExtensionResource extension_resources = 3;
}

// RouteExtensionContext provides context about an HTTPRoute/GRPCRoute containing an
// ExtensionRef. The context provided includes which filterchain/s
// and route of the object referent.
message RouteExtensionContext {
    // route_name is the name of HTTPRoute the filter is being applied on
    string route_name = 1;

    // route_namespace is the namespace the HTTPRoute is located in
    string route_namespace = 2;

    // hostnames are the fully qualified domain names attached to the HTTPRoute
    repeated string hostnames = 3;
}

// ExtensionResource stores the data for a K8s API object referenced in an HTTPRouteFilter
// extensionRef. It is constructed from an unstructured.Unstructured marshaled to JSON. An extension
// can marshal the bytes from this resource back into an unstructured.Unstructured and then 
// perform type checking to obtain the resource.
message ExtensionResource {

>     bytes unstructured_bytes = 1;

Reviewer 5: lets add some structure to this

// For example, an extension may want to write specific Typed Per-Filter Config for all routes.
message VirtualHostModifyRequest {
    envoy.config.route.v3.VirtualHost virtual_host = 1;
}

// VirtualHostModifyResponse contains the xDS VirtualHost that was generated by Envoy Gateway after it has been modified by
// an extension
message VirtualHostModifyResponse {
    envoy.config.route.v3.VirtualHost virtual_host = 1;
}
```

### HTTP Listener Modification Hook

The HTTP Listener modification hook is the broadest xDS modification Hook available and allows an extension to make many changes at a higher level. The HTTP Listener Hook allows an extension to make changes to a Listener after it was generated by Envoy Gateway. It is always executed when an extension is loaded. An extension may return an unmodified version of the Listener it
received in order to not make any changes to it, or return nil to cause the Listener to be discarded.

```protobuf
// HTTPListenerModifyRequest sends over the Listener that was generated by Envoy Gateway so that an extension can modify it
message HTTPListenerModifyRequest {

>     envoy.config.listener.v3.Listener listener = 1;

Reviewer 5: no context needed here ?

Extending Envoy Gateway by using an external extension server which makes use of hook points in Envoy Gateway does come with some known issues that will have to be worked around.

- Extension server handling its own state such as in the scenario where an extension establishes watches its own CRDs in the cluster can lead to a scenario where
ensuring that the extension and Envoy Gateway are in-sync becomes a challenge. There are two potential solutions to this problem.

>   - Option 1: Create a new External Provider mechanism where an extension can take over all provider responsibility for Envoy Gateway. The current functionality of the provider would need to become a Library that EG would make use of 

Reviewer 5: I think the option 1 suggested on the call today was - EG reconciles/watches only EG resources, and the extension watches the Extension resources

Reviewer 4: I agree this is a better way to seperates concerns. However, if keeping EG and Extensions in sync poses a problem, I prefer 'the provider listens to all resources and passes them back to the extension as needed' approach. I think we should put too much responsibility on the Extension side.
